src/selection.js

```diff
@@ -14,33 +14,72 @@ module.exports = Selection;
  * facilities to manipulate, save, restore the
  * selection, across DOM modifications.
  */
-function Selection(nativeSelection) {
+function Selection(nativeSelection, editor) {
   var me = this;
+  me.editor = editor;
 
   var native = me.native = assert(nativeSelection);
 
   me.setCaret = function(point) {
     var pair = Point.check(point).toNodeOffset();
 
-    native.setBaseAndExtent(pair[0], pair[1], pair[0], pair[1]);
+    me.setBaseAndExtent(pair[0], pair[1], pair[0], pair[1]);
   };
 
   me.setEndpoints = function(anchor, focus) {
     var anchorPair = Point.check(anchor).toNodeOffset();
     var focusPair = Point.check(focus).toNodeOffset();
 
-    native.setBaseAndExtent(anchorPair[0], anchorPair[1], focusPair[0], focusPair[1]);
+    me.setBaseAndExtent(anchorPair[0], anchorPair[1], focusPair[0], focusPair[1]);
   };
 
-  /** Convenience function - returns true if the selection is collapsed */
+  me.setBaseAndExtent = function(anchorParent, anchorOffset, focusParent, focusOffset) {
+    if (native.setBaseAndExtent) {
+      native.setBaseAndExtent(anchorParent, anchorOffset, focusParent, focusOffset);
+    } else {
+      // Using Firefox caret setting methods as backup
+      var fRange = native.getRangeAt(0);
+      fRange.setStart(anchorParent, anchorOffset);
+      fRange.setEnd(focusParent, focusOffset);
+
+    }
+  };
+
+  /** Convenience function - returns true if the selection is collapsed or there is no range*/
   me.isCollapsed = function() {
-    return me.getRange().isCollapsed();
+    var range = me.getRange();
+    if (range) {
+      return range.isCollapsed();
+    } else {
+      return true;
+    }
+  };
+
+  /** Helper function - returns true if the selection is within the editor elem */
+  me.withinEditor = function() {
+    if (!native.anchorNode || !native.focusNode) {
+      return false;
+    } else {
+      var within = false;
+
+      try {
+        within = util.compareNodes(me.editor.currentElem(), native.anchorNode) === "parent" &&
```

Comment:
FYI as cool as compareNodes is, we can probably replace its implementation with the native version that I believe most browsers now support (not urgent)